### PR TITLE
[SYCL] Make sycl::marray trivially copyable

### PR DESCRIPTION
**Pull request by psamolysov-intel**
The sycl::marray class is not trivially copyable but marked as device copyable. It makes problems with any class that contain a member with the sycl::marray type: that class will not be trivially copyable and therefore device copyable regardless the sycl::marray one is. To solve this issue, the sycl::marray class is converted into a trivially copyable one. The constexpr attribute for the copy constructor is saved so the class can still be copied in constant expressions.

### CODE REVIEW

**Comment by psamolysov-intel** [comment]
The changes break the ABI since the class will be marked as trivially copyable.

File: sycl/test/basic_tests/SYCL-2020-spec-constants.cpp [hunk 1]
```diff
@@ -38,6 +38,8 @@ constexpr sycl::specialization_id<uint64_t> uint64_id(8);
 constexpr sycl::specialization_id<float> float_id(10.0);
 constexpr sycl::specialization_id<double> double_id(11.0);
 constexpr sycl::marray<double, 5> ma;
+constexpr sycl::marray<double, 5> mb(ma);
```
Comment (vladimirlaz) [inline sycl/test/basic_tests/SYCL-2020-spec-constants.cpp:41]: It is not clear why these lines were added to the test?
tey are not used in specialization constants and this test is for specialization constants.

Comment (psamolysov-intel) [reply sycl/test/basic_tests/SYCL-2020-spec-constants.cpp]: Agree, I've moved the lines into `marray.cpp` test.

**Review by vladimirlaz — DISMISSED** [review]

**Comment by psamolysov-intel** [comment]
A test with a struct containing `sycl::marray` has been added into `sycl/test/basic_tests/valid_kernel_args.cpp`.

**Comment by Pennycook** [comment]
Should we also add a regression test here?  I'm thinking it would be good to check via a `static_assert` that:
- `marray<T>` is trivially copyable for some trivially copyable `T`
- `marray<T>` is device copyable for some device copyable `T`
- `marray<T>` is not trivially copyable for some non-trivially copyable `T`
- `marray<T>` is not device copyable for some non-device copyable `T`

**Comment by gmlueck** [comment]
I think these changes look good.  I agree with @Pennycook about adding some more tests.

In addition, we should probably add code to conditionally define the trait `is_device_copyable<marray<T>>` to `std::true` when `T` is device copyable.

File: sycl/include/CL/sycl/marray.hpp [hunk 2]
```diff
@@ -66,17 +66,9 @@ template <typename Type, std::size_t NumElements> class marray {
       typename = typename std::enable_if<sizeof...(ArgTN) == NumElements>::type>
   constexpr marray(const ArgTN &... Args) : MData{Args...} {}
 
-  constexpr marray(const marray<Type, NumElements> &Rhs) {
-    for (std::size_t I = 0; I < NumElements; ++I) {
-      MData[I] = Rhs.MData[I];
-    }
-  }
+  constexpr marray(const marray<Type, NumElements> &Rhs) = default;
```
Comment (keryell) [inline sycl/include/CL/sycl/marray.hpp:69]: I love C++! :-)

**Comment by psamolysov-intel** [comment]
@gmlueck Sorry, I don't understand why we should define the trait `is_device_copyable<marray<T>>` to `std::true` when `T` is device copyable if I removed that trait in the patch since it makes a conflict with our basic trait `is_device_copyable = true` for a trivially copyable `T`. If there are both traits, the compiler generates an error since there are ambiguous specializations of the `is_device_copyable` templates.

**Comment by psamolysov-intel** [comment]
@Pennycook Thank you for the suggestion. I've added the tests in `sycl\test\basic_tests\marray\marray.cpp`.

**Comment by psamolysov-intel** [comment]
`marray<tuple<>, ...>` is a bad example, since it considered trivially copyable on Linux with `libstdc++` and not considered on Windows with MSVC.

**Review by gmlueck — APPROVED** [review]

**Review by Pennycook — APPROVED** [review]

**Pull request merged**